File: gitdot-server/src/handler/migration/github/list_github_installation_repositories.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
};

use gitdot_api::endpoint::migration::github::list_github_installation_repositories as api;
use gitdot_core::dto::ListGitHubInstallationRepositoriesRequest;

use crate::{
    app::{AppError, AppResponse, AppState},
    dto::IntoApi,
    extract::{Principal, User},
};

#[axum::debug_handler]
pub async fn list_github_installation_repositories(
    auth_user: Principal<User>,
    State(state): State<AppState>,
    Path(installation_id): Path<i64>,
) -> Result<AppResponse<api::ListGitHubInstallationRepositoriesResponse>, AppError> {
    let request = ListGitHubInstallationRepositoriesRequest {
        owner_id: auth_user.id,
        installation_id,
    };
    state
        .migration_service
        .list_github_installation_repositories(request)
        .await
        .map_err(AppError::from)
        .map(|repos| AppResponse::new(StatusCode::OK, repos.into_api()))
}

File: gitdot-web/packages/gitdot-api/src/endpoint/migration/github/list-github-installation-repositories.ts
import { z } from "zod";
import { GitHubRepositoryResource } from "../../../resource";
import type { Endpoint } from "../../endpoint";

export const ListGitHubInstallationRepositoriesRequest = z.object({});
export type ListGitHubInstallationRepositoriesRequest = z.infer<
  typeof ListGitHubInstallationRepositoriesRequest
>;

export const ListGitHubInstallationRepositoriesResponse = z.array(
  GitHubRepositoryResource,
);
export type ListGitHubInstallationRepositoriesResponse = z.infer<
  typeof ListGitHubInstallationRepositoriesResponse
>;

export const ListGitHubInstallationRepositories = {
  path: "/migration/github/{installation_id}/repositories",
  method: "GET",
  request: ListGitHubInstallationRepositoriesRequest,
  response: ListGitHubInstallationRepositoriesResponse,
} as const satisfies Endpoint;
export type ListGitHubInstallationRepositories =
  typeof ListGitHubInstallationRepositories;
